File: python/diagram_printer.py
import os
import logging

from documents import DiagramSummary, FlowchartDiagram, DiagramMetadata
from physical_printer import DiagramPhysicalPrinter
from reporting import DiagramPagesReport, FlowchartReportItems, PagesBuilder


# This is a class you'd like to get under test so you can change it safely.
class DiagramPrinter:
    SPREADSHEET = "Spreadsheet"
    PDF = "PDF"

    def __init__(self):
        self._logger = logging.getLogger("DiagramPrinter")

    def print_summary(self, diagram, language):
        if diagram is None:
            return False, ""

        summary = DiagramSummary(language)
        summary.add_title(diagram.name(), diagram.serial_number())
        summary.add_header(diagram.summary_information())
        summary.add_image(diagram.flowchart_thumbnail())
        summary_text = summary.export()
        return True, summary_text

    def print_diagram(self, diagram: FlowchartDiagram, folder: str = None, filename: str = None):
        if diagram is None:
            return False

        info = DiagramMetadata(diagram)
        target_filename = self._get_target_filename(folder, filename)

        if info.file_type == self.PDF:
            self._logger.info(f"Printing PDF to file {target_filename}")
            return diagram.flowchart_as_pdf().copy_file(info.full_filename, target_filename, True)

        if info.file_type == self.SPREADSHEET:
            if not target_filename.endswith(".xls"):
                target_filename += ".xls"
            self._logger.info(f"Printing Excel to file {target_filename}")
            return diagram.flowchart_data_as_spreadsheet().copy_file(info.full_filename, target_filename, True)

        # default case - print to a physical printer
        diagram_physical_printer = DiagramPhysicalPrinter()
        return diagram_physical_printer.do_print(diagram, info, target_filename)

    def print_report(self, diagram: FlowchartDiagram, report_template: str, folder: str,
                     filename: str, summarize: bool) -> bool:
        if diagram is None:
            return False

        report = diagram.report()
        target_filename = self._get_target_filename(folder, filename)
        self._logger.info(f"Creating report for {diagram.name()} to file {target_filename}")

        if summarize:
            diagram = diagram.summary()
            report.close()
            report = diagram.report()
            report.open(True)
            self._logger.info(f"Switched to summary report for {diagram.name()}")

        if not report.is_open():
            self._logger.warning("Failed to open report for writing.")
            return False

        data = diagram.report_data()

        if not self._validate_report(report_template, data):
            self._logger.warning("Failed to validate report template.")
            return False

        if summarize:
            data.add(diagram.summary_information())
            report.open_with_contents(report_template, data, True)
        else:
            report.open_with_contents(report_template, data, False)

        report.save_to_file(target_filename)
        self._logger.info("Report creation succeeded")
        return True

    def print_pages(self, diagram: FlowchartDiagram, builder: 'PagesBuilder') -> bool:
        if diagram is None:
            return False

        data = diagram.report_data()
        pages = diagram.pages_data()

        report = DiagramPagesReport()
        report_pages = []

        for page in pages:
            processed_page = builder.process_page(page, data)
            report_pages.append(processed_page)

        return builder.apply(report, report_pages)

    @staticmethod
    def _get_target_filename(folder, filename):
        folder = folder or os.path.expanduser("~")
        filename = filename or "tempfile.tmp"
        return os.path.join(folder, filename)

    def _validate_report(self, template: str, substitutions: 'FlowchartReportItems') -> bool:
        try:
            self.create_report(template, substitutions.to_array())
            return True
        except ValueError as e:
            self._logger.warning("Report template did not match substitutions")
            return False

    def create_report(self, template: str, substitutions: list) -> str:
        return template.format(*substitutions)

File: typescript/src/diagram_printer.ts
import * as path from 'path';
import * as os from 'os';
import { DiagramSummary, FlowchartDiagram, DiagramMetadata } from './documents';
import { DiagramPhysicalPrinter } from './physical_printer';
import {DiagramPagesReport, DiagramReportPage, FlowchartReportItems, PagesBuilder} from "./reporting";
import * as util from "util";
import {PrintableDiagram} from "./printableDiagram";


class DiagramPrinter {
  static readonly SPREADSHEET = "Spreadsheet";
  static readonly PDF = "PDF";

  constructor() {
  }

  printSummary(diagram: FlowchartDiagram | null, language: string): [boolean, string] {
    if (diagram === null) {
      return [false, ""];
    }

    const summary = new DiagramSummary(language);
    summary.addTitle(diagram.name(), diagram.serialNumber());
    summary.addHeader(diagram.summaryInformation());
    summary.addImage(diagram.flowchartThumbnail());
    const summaryText = summary.export();
    return [true, summaryText];
  }

  async printDiagram(diagram: FlowchartDiagram | null, folder?: string, filename?: string): Promise<boolean> {
    if (diagram === null) {
      return false;
    }

    let printableDiagram = new PrintableDiagram(diagram);
    return this.printDiagramInternal(printableDiagram, folder, filename);
  }

  public async printDiagramInternal(
    printableDiagram: PrintableDiagram,
    folder?: string,
    filename?: string
  ): Promise<boolean> {
    const info = printableDiagram.getDiagramMetadata();

    if (info.fileType === DiagramPrinter.PDF) {
      const targetFilename = this._getTargetFilename(folder, filename);
      return printableDiagram.printToFile(info.fullFilename, targetFilename);
    }

    if (info.fileType === DiagramPrinter.SPREADSHEET) {
      let targetFilename = this._getTargetFilename(folder, filename);
      if (!targetFilename.endsWith(".xls")) {
        targetFilename += ".xls";
      }
      return printableDiagram.printToSpreadsheetFile(info.fullFilename, targetFilename);
    }

    // Default case - print to a physical printer
    const physicalPrinter = new DiagramPhysicalPrinter();
    return await physicalPrinter.doPrint(printableDiagram, info, this._getTargetFilename(folder, filename));
  }


  printReport(
    diagram: FlowchartDiagram | null,
    reportTemplate: string,
    folder: string,
    filename: string,
    summarize: boolean
  ): boolean {
    if (diagram === null) {
      return false;
    }

    let report = diagram.report();
    const targetFilename = this._getTargetFilename(folder, filename);
    console.info(`Creating report for ${diagram.name()} to file ${targetFilename}`);

    if (summarize) {
      diagram = diagram.summary();
      report.close();
      report = diagram.report();
      report.open(true);
      console.info(`Switched to summary report for ${diagram.name()}`);
    }

    if (!report.isOpen()) {
      console.error("Failed to open report for writing.");
      return false;
    }

    const data = diagram.reportData();

    if (!this.validateReport(reportTemplate, data)) {
      console.error("Failed to validate report template.");
      return false;
    }

    if (summarize) {
      data.add(diagram.summaryInformation());
      report.openWithContents(reportTemplate, data, true);
    } else {
      report.openWithContents(reportTemplate, data, false);
    }

    report.saveToFile(targetFilename);
    console.info("Report creation succeeded");
    return true;
  }

  printPages(diagram: FlowchartDiagram | null, builder: PagesBuilder): boolean {
    if (diagram === null) {
      return false;
    }

    const data = diagram.reportData();
    const pages = diagram.pagesData();

    const report = new DiagramPagesReport();
    const reportPages: DiagramReportPage[] = [];

    for (const page of pages) {
      const processedPage = builder.processPage(page, data);
      reportPages.push(processedPage);
    }

    return builder.apply(report, reportPages);
  }

  private _getTargetFilename(folder?: string, filename?: string): string {
    const targetFolder = folder || os.homedir();
    const targetFilename = filename || "tempfile.tmp";
    return path.join(targetFolder, targetFilename);
  }

  validateReport(template: string, substitutions: FlowchartReportItems): boolean {
    try {
      this.createReport(template, substitutions.toArray());
      return true;
    } catch (error) {
      console.error("Report template did not match substitutions");
      return false;
    }
  }

  createReport(template: string, substitutions: any[]): string {
    let report = util.format(template, ...substitutions);
    return report;
  }
}

export { DiagramPrinter };
